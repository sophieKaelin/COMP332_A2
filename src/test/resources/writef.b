//
// This file is part of COMP332 Assignment 2/3 2019.
//
// weBCPL, a retro BCPL to WebAssembly compiler.
//
// © 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
// © 2005, Martin Richards
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This code implements the BCPL `writef` function, which evolved into the
// `printf` function in the C standard library. It is taken from the original 
// BCPL standard library written by Martin Richards.
//
// As this stands this code will only work with those weBCPL strings that
// confine themselves to the ASCII portion of the UTF-8 encoding.
//

LET writed(n, d) BE
$(
    LET t = VEC 30
    AND i, k = 0, -n
    IF n<0 DO d, k := d-1, n
    t!i, i, k := -(k REM 10), i+1, k/10 REPEATUNTIL k=0
    FOR j = i+1 TO d DO wrch('*s')
    IF n<0 DO wrch('-')
    FOR j = i-1 TO 0 BY -1 DO wrch(t!j+'0')
$)

AND writeu(n, d) BE
$(
    LET m = (n>>1)/5
    UNLESS m=0 DO $( writed(m, d-1); d := 1 $)
    writed(n-m*10, d)
$)

AND writen(n) BE writed(n, 0)

AND writebin(n, d) BE
$(
    IF d>1 DO writebin(n>>1, d-1)
    wrch((n&1)+'0')
$)

AND writeoct(n, d) BE
$(
    IF d>1 DO writeoct(n>>3, d-1)
    wrch((n&7)+'0')
$)

AND writehex(n, d) BE
$(
    IF d>1 DO writehex(n>>4, d-1)
    wrch((TABLE '0','1','2','3','4','5','6','7',
                '8','9','A','B','C','D','E','F')!(n&15) )
$)

AND writes(s) BE FOR i = 1 TO s%0 DO wrch(s%i)

AND writet(s, d) BE
$(
    writes(s)
    FOR i = 1 TO d-s%0 DO wrch('*s')
$)

AND writef(format, a, b, c, d, e, f, g, h, i, j, k) BE
$(
    LET t = @ a
    FOR p = 1 TO format%0 DO
    $(
        LET k = format%p
        TEST k='%' THEN 
        $(
            LET f, n = ?, ?
            p := p+1
            SWITCHON capitalch(format%p) INTO
            $(
                DEFAULT:  wrch(format%p); ENDCASE
                CASE 'S': f := writes;    GOTO l
                CASE 'T': f := writet;    GOTO m
                CASE 'C': f := wrch;      GOTO l
                CASE 'B': f := writebin;  GOTO m
                CASE 'O': f := writeoct;  GOTO m
                CASE 'X': f := writehex;  GOTO m
                CASE 'I': f := writed;    GOTO m
                CASE 'N': f := writen;    GOTO l
                CASE 'U': f := writeu
                m:        p := p+1
                          n := format%p
                          n := '0'<=n<='9' -> n-'0', 10+n-'A'
                l:        f(!t, n)
                CASE '$': t := t+1
            $)
        $)
        ELSE wrch(k)
    $)
$)
