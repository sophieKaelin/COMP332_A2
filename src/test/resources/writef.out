Program(
    Vector(
        LetDecl(
            Vector(
                LetProcClause(
                    IdnDef("writed"),
                    Vector(IdnDef("n"), IdnDef("d")),
                    Block(
                        Vector(
                            LetDecl(
                                Vector(
                                    LetVecClause(IdnDef("t"), IntExp(30)),
                                    LetVarClause(
                                        Vector(IdnDef("i"), IdnDef("k")),
                                        Vector(IntExp(0), NegExp(IdnExp(IdnUse("n")))))))),
                        Vector(
                            IfDoStmt(
                                LessExp(IdnExp(IdnUse("n")), IntExp(0)),
                                AssignStmt(
                                    Vector(IdnExp(IdnUse("d")), IdnExp(IdnUse("k"))),
                                    Vector(
                                        MinusExp(IdnExp(IdnUse("d")), IntExp(1)),
                                        IdnExp(IdnUse("n"))))),
                            RepeatUntilStmt(
                                AssignStmt(
                                    Vector(
                                        BinaryPlingExp(IdnExp(IdnUse("t")), IdnExp(IdnUse("i"))),
                                        IdnExp(IdnUse("i")),
                                        IdnExp(IdnUse("k"))),
                                    Vector(
                                        NegExp(ModExp(IdnExp(IdnUse("k")), IntExp(10))),
                                        PlusExp(IdnExp(IdnUse("i")), IntExp(1)),
                                        SlashExp(IdnExp(IdnUse("k")), IntExp(10)))),
                                EqualExp(IdnExp(IdnUse("k")), IntExp(0))),
                            ForStmt(
                                IdnDef("j"),
                                PlusExp(IdnExp(IdnUse("i")), IntExp(1)),
                                IdnExp(IdnUse("d")),
                                None,
                                CallStmt(CallExp(IdnExp(IdnUse("wrch")), Vector(ChrExp(Vector(32)))))),
                            IfDoStmt(
                                LessExp(IdnExp(IdnUse("n")), IntExp(0)),
                                CallStmt(CallExp(IdnExp(IdnUse("wrch")), Vector(ChrExp(Vector(45)))))),
                            ForStmt(
                                IdnDef("j"),
                                MinusExp(IdnExp(IdnUse("i")), IntExp(1)),
                                IntExp(0),
                                Some(NegExp(IntExp(1))),
                                CallStmt(
                                    CallExp(
                                        IdnExp(IdnUse("wrch")),
                                        Vector(
                                            PlusExp(
                                                BinaryPlingExp(
                                                    IdnExp(IdnUse("t")),
                                                    IdnExp(IdnUse("j"))),
                                                ChrExp(Vector(48)))))))))),
                LetProcClause(
                    IdnDef("writeu"),
                    Vector(IdnDef("n"), IdnDef("d")),
                    Block(
                        Vector(
                            LetDecl(
                                Vector(
                                    LetVarClause(
                                        Vector(IdnDef("m")),
                                        Vector(
                                            SlashExp(
                                                ShiftRightExp(IdnExp(IdnUse("n")), IntExp(1)),
                                                IntExp(5))))))),
                        Vector(
                            UnlessDoStmt(
                                EqualExp(IdnExp(IdnUse("m")), IntExp(0)),
                                Block(
                                    Vector(),
                                    Vector(
                                        CallStmt(
                                            CallExp(
                                                IdnExp(IdnUse("writed")),
                                                Vector(
                                                    IdnExp(IdnUse("m")),
                                                    MinusExp(IdnExp(IdnUse("d")), IntExp(1))))),
                                        AssignStmt(Vector(IdnExp(IdnUse("d"))), Vector(IntExp(1)))))),
                            CallStmt(
                                CallExp(
                                    IdnExp(IdnUse("writed")),
                                    Vector(
                                        MinusExp(
                                            IdnExp(IdnUse("n")),
                                            StarExp(IdnExp(IdnUse("m")), IntExp(10))),
                                        IdnExp(IdnUse("d")))))))),
                LetProcClause(
                    IdnDef("writen"),
                    Vector(IdnDef("n")),
                    CallStmt(
                        CallExp(IdnExp(IdnUse("writed")), Vector(IdnExp(IdnUse("n")), IntExp(0))))),
                LetProcClause(
                    IdnDef("writebin"),
                    Vector(IdnDef("n"), IdnDef("d")),
                    Block(
                        Vector(),
                        Vector(
                            IfDoStmt(
                                GreaterExp(IdnExp(IdnUse("d")), IntExp(1)),
                                CallStmt(
                                    CallExp(
                                        IdnExp(IdnUse("writebin")),
                                        Vector(
                                            ShiftRightExp(IdnExp(IdnUse("n")), IntExp(1)),
                                            MinusExp(IdnExp(IdnUse("d")), IntExp(1)))))),
                            CallStmt(
                                CallExp(
                                    IdnExp(IdnUse("wrch")),
                                    Vector(
                                        PlusExp(
                                            AndExp(IdnExp(IdnUse("n")), IntExp(1)),
                                            ChrExp(Vector(48))))))))),
                LetProcClause(
                    IdnDef("writeoct"),
                    Vector(IdnDef("n"), IdnDef("d")),
                    Block(
                        Vector(),
                        Vector(
                            IfDoStmt(
                                GreaterExp(IdnExp(IdnUse("d")), IntExp(1)),
                                CallStmt(
                                    CallExp(
                                        IdnExp(IdnUse("writeoct")),
                                        Vector(
                                            ShiftRightExp(IdnExp(IdnUse("n")), IntExp(3)),
                                            MinusExp(IdnExp(IdnUse("d")), IntExp(1)))))),
                            CallStmt(
                                CallExp(
                                    IdnExp(IdnUse("wrch")),
                                    Vector(
                                        PlusExp(
                                            AndExp(IdnExp(IdnUse("n")), IntExp(7)),
                                            ChrExp(Vector(48))))))))),
                LetProcClause(
                    IdnDef("writehex"),
                    Vector(IdnDef("n"), IdnDef("d")),
                    Block(
                        Vector(),
                        Vector(
                            IfDoStmt(
                                GreaterExp(IdnExp(IdnUse("d")), IntExp(1)),
                                CallStmt(
                                    CallExp(
                                        IdnExp(IdnUse("writehex")),
                                        Vector(
                                            ShiftRightExp(IdnExp(IdnUse("n")), IntExp(4)),
                                            MinusExp(IdnExp(IdnUse("d")), IntExp(1)))))),
                            CallStmt(
                                CallExp(
                                    IdnExp(IdnUse("wrch")),
                                    Vector(
                                        BinaryPlingExp(
                                            TableExp(
                                                Vector(
                                                    ChrExp(Vector(48)),
                                                    ChrExp(Vector(49)),
                                                    ChrExp(Vector(50)),
                                                    ChrExp(Vector(51)),
                                                    ChrExp(Vector(52)),
                                                    ChrExp(Vector(53)),
                                                    ChrExp(Vector(54)),
                                                    ChrExp(Vector(55)),
                                                    ChrExp(Vector(56)),
                                                    ChrExp(Vector(57)),
                                                    ChrExp(Vector(65)),
                                                    ChrExp(Vector(66)),
                                                    ChrExp(Vector(67)),
                                                    ChrExp(Vector(68)),
                                                    ChrExp(Vector(69)),
                                                    ChrExp(Vector(70)))),
                                            AndExp(IdnExp(IdnUse("n")), IntExp(15))))))))),
                LetProcClause(
                    IdnDef("writes"),
                    Vector(IdnDef("s")),
                    ForStmt(
                        IdnDef("i"),
                        IntExp(1),
                        BinaryBytePlingExp(IdnExp(IdnUse("s")), IntExp(0)),
                        None,
                        CallStmt(
                            CallExp(
                                IdnExp(IdnUse("wrch")),
                                Vector(BinaryBytePlingExp(IdnExp(IdnUse("s")), IdnExp(IdnUse("i")))))))),
                LetProcClause(
                    IdnDef("writet"),
                    Vector(IdnDef("s"), IdnDef("d")),
                    Block(
                        Vector(),
                        Vector(
                            CallStmt(CallExp(IdnExp(IdnUse("writes")), Vector(IdnExp(IdnUse("s"))))),
                            ForStmt(
                                IdnDef("i"),
                                IntExp(1),
                                MinusExp(
                                    IdnExp(IdnUse("d")),
                                    BinaryBytePlingExp(IdnExp(IdnUse("s")), IntExp(0))),
                                None,
                                CallStmt(CallExp(IdnExp(IdnUse("wrch")), Vector(ChrExp(Vector(32))))))))),
                LetProcClause(
                    IdnDef("writef"),
                    Vector(
                        IdnDef("format"),
                        IdnDef("a"),
                        IdnDef("b"),
                        IdnDef("c"),
                        IdnDef("d"),
                        IdnDef("e"),
                        IdnDef("f"),
                        IdnDef("g"),
                        IdnDef("h"),
                        IdnDef("i"),
                        IdnDef("j"),
                        IdnDef("k")),
                    Block(
                        Vector(
                            LetDecl(
                                Vector(
                                    LetVarClause(
                                        Vector(IdnDef("t")),
                                        Vector(AddrOfExp(IdnExp(IdnUse("a")))))))),
                        Vector(
                            ForStmt(
                                IdnDef("p"),
                                IntExp(1),
                                BinaryBytePlingExp(IdnExp(IdnUse("format")), IntExp(0)),
                                None,
                                Block(
                                    Vector(
                                        LetDecl(
                                            Vector(
                                                LetVarClause(
                                                    Vector(IdnDef("k")),
                                                    Vector(
                                                        BinaryBytePlingExp(
                                                            IdnExp(IdnUse("format")),
                                                            IdnExp(IdnUse("p")))))))),
                                    Vector(
                                        TestThenElseStmt(
                                            EqualExp(IdnExp(IdnUse("k")), ChrExp(Vector(37))),
                                            Block(
                                                Vector(
                                                    LetDecl(
                                                        Vector(
                                                            LetVarClause(
                                                                Vector(IdnDef("f"), IdnDef("n")),
                                                                Vector(UndefExp(), UndefExp()))))),
                                                Vector(
                                                    AssignStmt(
                                                        Vector(IdnExp(IdnUse("p"))),
                                                        Vector(
                                                            PlusExp(IdnExp(IdnUse("p")), IntExp(1)))),
                                                    SwitchOnStmt(
                                                        CallExp(
                                                            IdnExp(IdnUse("capitalch")),
                                                            Vector(
                                                                BinaryBytePlingExp(
                                                                    IdnExp(IdnUse("format")),
                                                                    IdnExp(IdnUse("p"))))),
                                                        Block(
                                                            Vector(),
                                                            Vector(
                                                                Default(
                                                                    CallStmt(
                                                                        CallExp(
                                                                            IdnExp(IdnUse("wrch")),
                                                                            Vector(
                                                                                BinaryBytePlingExp(
                                                                                    IdnExp(
                                                                                        IdnUse(
                                                                                            "format")),
                                                                                    IdnExp(
                                                                                        IdnUse("p"))))))),
                                                                EndCaseStmt(),
                                                                CaseOf(
                                                                    ChrExp(Vector(83)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(IdnUse("writes"))))),
                                                                GotoStmt(LabUse("l")),
                                                                CaseOf(
                                                                    ChrExp(Vector(84)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(IdnUse("writet"))))),
                                                                GotoStmt(LabUse("m")),
                                                                CaseOf(
                                                                    ChrExp(Vector(67)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(IdnUse("wrch"))))),
                                                                GotoStmt(LabUse("l")),
                                                                CaseOf(
                                                                    ChrExp(Vector(66)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(
                                                                                IdnUse("writebin"))))),
                                                                GotoStmt(LabUse("m")),
                                                                CaseOf(
                                                                    ChrExp(Vector(79)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(
                                                                                IdnUse("writeoct"))))),
                                                                GotoStmt(LabUse("m")),
                                                                CaseOf(
                                                                    ChrExp(Vector(88)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(
                                                                                IdnUse("writehex"))))),
                                                                GotoStmt(LabUse("m")),
                                                                CaseOf(
                                                                    ChrExp(Vector(73)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(IdnUse("writed"))))),
                                                                GotoStmt(LabUse("m")),
                                                                CaseOf(
                                                                    ChrExp(Vector(78)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(IdnUse("writen"))))),
                                                                GotoStmt(LabUse("l")),
                                                                CaseOf(
                                                                    ChrExp(Vector(85)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("f"))),
                                                                        Vector(
                                                                            IdnExp(IdnUse("writeu"))))),
                                                                Labelled(
                                                                    LabDef("m"),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("p"))),
                                                                        Vector(
                                                                            PlusExp(
                                                                                IdnExp(IdnUse("p")),
                                                                                IntExp(1))))),
                                                                AssignStmt(
                                                                    Vector(IdnExp(IdnUse("n"))),
                                                                    Vector(
                                                                        BinaryBytePlingExp(
                                                                            IdnExp(IdnUse("format")),
                                                                            IdnExp(IdnUse("p"))))),
                                                                AssignStmt(
                                                                    Vector(IdnExp(IdnUse("n"))),
                                                                    Vector(
                                                                        IfExp(
                                                                            AndExp(
                                                                                LessOrEqualExp(
                                                                                    ChrExp(
                                                                                        Vector(48)),
                                                                                    IdnExp(
                                                                                        IdnUse("n"))),
                                                                                LessOrEqualExp(
                                                                                    IdnExp(
                                                                                        IdnUse("n")),
                                                                                    ChrExp(
                                                                                        Vector(57)))),
                                                                            MinusExp(
                                                                                IdnExp(IdnUse("n")),
                                                                                ChrExp(Vector(48))),
                                                                            MinusExp(
                                                                                PlusExp(
                                                                                    IntExp(10),
                                                                                    IdnExp(
                                                                                        IdnUse("n"))),
                                                                                ChrExp(Vector(65)))))),
                                                                Labelled(
                                                                    LabDef("l"),
                                                                    CallStmt(
                                                                        CallExp(
                                                                            IdnExp(IdnUse("f")),
                                                                            Vector(
                                                                                UnaryPlingExp(
                                                                                    IdnExp(
                                                                                        IdnUse("t"))),
                                                                                IdnExp(IdnUse("n")))))),
                                                                CaseOf(
                                                                    ChrExp(Vector(36)),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("t"))),
                                                                        Vector(
                                                                            PlusExp(
                                                                                IdnExp(IdnUse("t")),
                                                                                IntExp(1)))))))))),
                                            CallStmt(
                                                CallExp(
                                                    IdnExp(IdnUse("wrch")),
                                                    Vector(IdnExp(IdnUse("k"))))))))))))))))
